databuilder/databuilder/extractor/query/json_query_metadata_extractor.py

```diff
@@ -0,0 +1,96 @@
+# Copyright Contributors to the Amundsen project.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from typing import (
+    Any, Dict, Iterator, List,
+)
+
+from databuilder.extractor.query.query_metadata_base import QueryMetadataBaseExtractor
+
+LOGGER = logging.getLogger(__name__)
+
+
+class JsonQueryMetadataExtractor(QueryMetadataBaseExtractor):
```

Comment:
any reason the input list needs to be in this specific format?

normally you will get a list of queries from query log, i wonder who did the transformation into this format?

Reply:
The thought here is that the objects in the list are relatively synonymous with the inputs required for each of the corresponding models and, for any implementer that _does_ end up parsing their own query, each of these objects can be extracted from a single query.

For example, the top level objects are:
- sql: straight forward, needs to come from _somewhere_ ( a file, query log, etc.)
- clean_sql (optional): a user defined sql that has been cleaned 
- user (optional): the user that executed the query
- query_tables (optional): corresponds to the model `QueryMetadata`
- query_wheres (optional): corresponds to the model `QueryWhereMetadata`
- query_joins (optional): this corresponds to the model `QueryJoinMetadata`

When an implementer subclasses `QueryMetadataBaseExtractor` they will have to implement `parse_query_record` at which point they _should_ be able to map their parsed query to the Amundsen parsed query.

Reply:
thanks for sharing, but out of curious : any reason we don't do it in the transformer side which is the typical way on how we transform  / decorate a message into the target form ?

Reply:
The only reason reason that a transformer wasn't used for the `JsonQueryMetadataExtractor` was because it was easier to set up an example JSON file that matched this exact structure and JSON was just easier than any other format that where there was a lower degree of cardinality between a Query and Joins / Where (which needs 1 to many).

I would expect that a custom implemented extractor would likely use a transformer inside of the `parse_query_record` function to both parse out their query and create the proper structure.

Reply:
hey @feng-tao just wanted to bump this when you have a moment - any other questions / feedback on this PR that you'd like addressed before approving?

Reply:
will take a closer look at the pr today, thanks.